c/tests/test_tables.c

```diff
@@ -1490,6 +1490,98 @@ test_node_table_update_row(void)
     tsk_node_table_free(&table);
 }
 
+static void
+test_node_table_keep_rows(void)
+{
+    int ret;
+    tsk_id_t ret_id;
+    tsk_size_t j;
+    tsk_node_table_t source, t1, t2;
+    tsk_node_t row;
+    bool keep[3] = { 1, 1, 1 };
+    tsk_id_t id_map[3];
+    const char *metadata = "ABC";
+
+    ret = tsk_node_table_init(&source, 0);
+    CU_ASSERT_EQUAL_FATAL(ret, 0);
+
+    ret_id = tsk_node_table_add_row(&source, 0, 1.0, 2, 3, metadata, 1);
+    CU_ASSERT_FATAL(ret_id >= 0);
+    ret_id = tsk_node_table_add_row(&source, 1, 2.0, 3, 4, metadata, 2);
+    CU_ASSERT_FATAL(ret_id >= 0);
+    ret_id = tsk_node_table_add_row(&source, 2, 3.0, 4, 5, metadata, 3);
+    CU_ASSERT_FATAL(ret_id >= 0);
+
+    ret = tsk_node_table_copy(&source, &t1, 0);
+    CU_ASSERT_EQUAL_FATAL(ret, 0);
+
+    ret = tsk_node_table_keep_rows(&t1, keep, 0, id_map);
+    CU_ASSERT_EQUAL_FATAL(ret, 0);
+    CU_ASSERT_TRUE(tsk_node_table_equals(&t1, &source, 0));
+
+    ret = tsk_node_table_keep_rows(&t1, keep, 0, NULL);
+    CU_ASSERT_EQUAL_FATAL(ret, 0);
+    CU_ASSERT_TRUE(tsk_node_table_equals(&t1, &source, 0));
+    CU_ASSERT_EQUAL_FATAL(id_map[0], 0);
+    CU_ASSERT_EQUAL_FATAL(id_map[1], 1);
+    CU_ASSERT_EQUAL_FATAL(id_map[2], 2);
+
+    keep[0] = 0;
+    keep[1] = 0;
+    keep[2] = 0;
+    ret = tsk_node_table_keep_rows(&t1, keep, 0, id_map);
+    CU_ASSERT_EQUAL_FATAL(ret, 0);
+    CU_ASSERT_EQUAL_FATAL(t1.num_rows, 0);
+    CU_ASSERT_EQUAL_FATAL(id_map[0], -1);
```

Comment:
I know it is a small thing, but I'm never quite sure where the line is between using `-1` and `TSK_NULL`. In the implementation it is `TSK_NULL`

Reply:
It's tricky all right, they're sort of interchangable in my head. You do need to know that TSK_NULL is -1 as a developer so I guess I'm just lazy with typing. Happy to change if you think it improves readability

Reply:
> It's tricky all right, they're sort of interchangable in my head. You do need to know that TSK_NULL is -1 as a developer so I guess I'm just lazy with typing. Happy to change if you think it improves readability

I'm curious about this too.  Why does one need to know that it is -1, as a developer?  Couldn't there be a strict rule that TSK_NULL should *always* be used, in cases where it applies?  Might it be helpful to actually change TSK_NULL to -2, for a test build, and see what breaks, and fix all of those places?  Etc.  I have always found this to be one of the more confusing parts of tskit's APIs; maybe it could be made more rigorous?  Anyhow, just random thoughts as I'm looking at these GitHub emails.  :->

Reply:
In this case you need to know that -1 is valid input and -2 isn't because we're testing the validation code. Given this knowledge, -1 is sometimes more readable than TSK_NULL (say, if we're declaring an array like ``{-1, 0, -1, 2}``). But it's a judgement call, and opinions will differ.

Reply:
Hmm.  To me, if -1 means `TSK_NULL` in a given context, then declaring an array as `{-1, 0, -1, 2}` would be very confusing, because it would imply that -1 is "just another value" and mask its special meaning, whereas `{TSK_NULL, 0, TSK_NULL, 2}` would be much more clear.  But as you say, I guess opinions vary!  :->  Coming from the C++ world, I would be itching to find a way to make using -1 be literally illegal and cause a compile error, in such contexts.  (I don't think such a mechanism exists, though, even in C++; things like "enum class" are gestures in that direction, but don't go far enough to cover the usage case that `TSK_NULL` inhabits.)  This is parallel to my opinion that using `0` to mean `NULL`/`nullptr` (in C/C++) is confusing and wrong; in C you don't get a compile error, but in C++ you actually do (or perhaps it's a warning, depending on settings?); `nullptr` was specifically invented to allow using 0 to become a compile error, as I understand it.  Anyhow, interesting discussion.  Personally, I'd vote for doing a pass over tskit to enforce API purity concerning `TSK_NULL`, but I happily admit that that is just my opinion!  :->

Reply:
(And I'd gently suggest that if this point is confusing even for @benjeffery, then it is probably a point of confusion worth clearing up for pretty much every user of tskit...)

Reply:
Well, I'm not sure I made this clear at the top, but I'm talking about tskit C library core developers here, not people writing client code. They are certainly not expected to know that TSK_NULL is -1. These tests are for core developers to read and work with. I don't think Ben is confused about -1 being TSK_NULL, he just wasn't sure when to use one or the other stylistically.

I really don't think we should enforce always using TSK_NULL within the library code, as this would result in code like this:
```c
if (id < TSK_NULL || id >= self->num_rows) {
    ..  error
}
```
You have to know that TSK_NULL is -1 for this to be safe, and given that, it's better to just write -1 and be explicit about it.

Or I guess you could do 
```c
if ((id != TSK_NULL && id < 0) || id >= self->num_rows) {
     ... error
}
```
then you could have TSK_NULL as -2, but what's the point? Writing 

```c
if (id < -1|| id >= self->num_rows) {
     ... error
}
```
is simple, clear and obvious that you're not allowing any loopholes.

TSK_NULL **will always be** -1. There are good reasons for it to be this value, from multiple different angles and I can foresee no circumstances in which it would be a good idea to change it to a different value.

So, gently disagreeing with suggestion :stuck_out_tongue_winking_eye:

Reply:
To me,

`if ((id < 0 || id >= self->num_rows) && (id != TSK_NULL))`

is clearer in its meaning.  The goal is not, of course, to redefine `TSK_NULL` to be -2 instead.  It's just to make it clear in the code that `TSK_NULL` is a special value.  With

`if (id < -1 || id >= self->num_rows)`

it looks, to my eye, as if either (a) there is a bug in the code and it allows -1 incorrectly, or (b) an index of -1 is for some reason permitted.  When written that way, the code does not express that the reason that -1 is permitted is that -1 is `TSK_NULL`; the reader has to _guess_ that that might be the reason, and is left wondering whether that guess is correct.

But I guess this comes down to personal coding style, and there's no need to beat the dead horse further.  :->  It's just interesting to explore these philosophical differences, I think.  I certainly understand better now why the tskit code treats -1 vs. `TSK_NULL` in the way it does.

Reply:
I think part of it is that -1 predates TSK_NULL usage by quite a long way, so (a) I'm used to thinking about it this way, and (b) lots of the code was written before TSK_NULL existed. I can see how it looks sloppy when coming from the other perspective of (correctly) thinking about TSK_NULL as an anonymous special value.